# 错误边界

默认情况下，若一个组件在渲染期间(render)发生错误，会导致整个组件数全部被卸载

错误边界:本质上就是一个组件，该组件会捕获到渲染期间(render)子组件发生的错误，并且有能力阻止错误继续传播

**ErrorBound**

**让某个组件处理捕获错误**

1. 编写生命周期函数 ```getDerivedStateFromError(err)```
    - 静态函数
    - 运行时间点：渲染子组件渲染子组件的过程中，发生错误之后，在更新页面之前，也就是说只有子组件发生错误，才会运行该函数
    - 该函数返回一个对象，React会将该对象的属性覆盖掉当前组件的state
    - 通常使用该函数来改变状态
    - 参数为错误对象

2. 编写生命周期函数 ```componentDidCatch```
    - 实例方法
    - 运行时间点：渲染子组件的过程中，发生错误，更新页面之后，由于其运行时间点比较靠后，因此不太会在该函数中改变状态
    - 这里运行的时候其实组件已经被卸载了
    - 通常，该函数用于记录错误消息(使用第二个参数来发送错误信息)
    - 参数为错误对象或者是info
```js
export default class ErrorBound extends PureComponent {

    state = {
        hasError: false
    }

    static getDerivedStateFromError(error) {
        console.log("发生错误了");
        return {
            hasError: true
        }
    }

    componentDidCatch(error, info) {
        console.log("记录错误信息");
        this.setState({})
    }



    render() {
        // setTimeout(() => {
        //     throw new Error("asfasdfasf");
        // }, 1000);
        if (this.state.hasError) {
            return <h1>出现错误了！</h1>
        }
        return this.props.children
    }
}
```

**细节**

某些错误，错误边界组件无法捕获

* 自己的组件里面的错误不会捕获，只能捕获子组件中的错误
* 异步错误也不会被捕获
* 事件中的错误

渲染的整个过程是同步的

总结：只会处理渲染子组件期间的同步错误